tools/azure-sdk-tools/devtools_testutils/proxy_startup.py

```diff
@@ -72,8 +73,26 @@ def get_container_info() -> "Optional[dict]":
 
     output, stderr = proc.communicate()
     try:
-        # This will succeed if we found a container with CONTAINER_NAME
-        return json.loads(output)
+        container_spec = json.loads(output)
+        if sys.platform.startswith("win") and os.environ.get("TF_BUILD"):
+            image_prefix = WINDOWS_IMAGE_SOURCE_PREFIX
+        # elif sys.platform.startswith("darwin"):
+        #     image_prefix = MAC_IMAGE_SOURCE_PREFIX
+        else:
+            image_prefix = LINUX_IMAGE_SOURCE_PREFIX
+
+        if container_spec['Image'] == image_prefix + ':' + get_image_tag():
+            # This will succeed if we found a container with CONTAINER_NAME
+            return container_spec
+        else:
+            remove_proc = subprocess.Popen(
+                shlex.split("docker rm -f " + CONTAINER_NAME),
```

Comment:
This line got me thinking about whether it's actually necessary to keep our test proxy container at all. The bulk of the work when starting a container is fetching the image if we don't have it locally -- but creating a container from a cached image takes almost no time.

I looked at [JavaScript's test proxy startup/teardown](https://github.com/Azure/azure-sdk-for-js/blob/main/common/tools/dev-tool/src/util/testProxyUtils.ts) code for inspiration, and they take the temporary container approach by starting up the container with a [`docker run --rm` command](https://github.com/Azure/azure-sdk-for-js/blob/1d1b9c44ac0d32af9ac9a348205cc6461c654540/common/tools/dev-tool/src/util/testProxyUtils.ts#L62) so that the container is deleted after tests run. The _image_ is cached locally still, so running tests doesn't take noticeably longer with this approach compared to what we're doing today.

I think we should do the same in Python by removing the `get_container_info` method and using `create_container` to create a new container whenever tests are run with Docker. The changes necessary are fairly significant, so I created a gist [here](https://gist.github.com/mccoyp/48fc6b262c2cab532f47aaf1d7882dcb) with the code. What do you think?

Reply:
This is exactly what I wanted. But I found one potential problem. The container is not deleted normally and remains due to a container creation error. The following is an example of an error caused by a port problem.
1. First, activate port 5000. I am using Mac, I use port 5000 by turning on the airplay function in the system. Maybe you can activate port 5000 in another container and get the same result.  

|Before|After|
|:-----:|:----:|
|<img width="809" alt="Screen Shot 2022-08-18 at 11 25 36 AM" src="https://user-images.githubusercontent.com/32767165/185279149-0f204dbf-11c5-47da-a832-fae6c6eeb80f.png">|<img width="809" alt="Screen Shot 2022-08-18 at 11 24 12 AM" src="https://user-images.githubusercontent.com/32767165/185279202-f1f111a1-1330-42b2-bcd6-029c03304a6d.png">|

2. Run the test proxy. It does not proceed due to an error.

3. You can see that the container was not deleted because it was not shut down properly due to an error.
<img width="1389" alt="Screen Shot 2022-08-18 at 11 17 40 AM" src="https://user-images.githubusercontent.com/32767165/185278245-7ee72e03-5ad7-4b6c-b439-e218d5d910a9.png">

4. In this state, stop the port 5000 and run the test proxy, but it does not work properly.
<img width="809" alt="Screen Shot 2022-08-18 at 11 29 00 AM" src="https://user-images.githubusercontent.com/32767165/185279601-4aadceaa-622c-4e17-866d-18ba6a596621.png">

<img width="816" alt="Screen Shot 2022-08-18 at 11 20 22 AM" src="https://user-images.githubusercontent.com/32767165/185278525-b0f94201-9f7a-4ddc-b5cc-054bcc4edfed.png">

It may be manually controlled by docker commands or desktops, but for more convenient use, I think we need code to see if containers exist. Therefore, i propose to create the delete_container() function and add it to the beginning of the start_container() function.

Please check the [code](https://gist.github.com/ponopono0322/3b9f8189f25bb36bbd64ecba4bf809ce) I suggest.

Reply:
Awesome, that makes sense to me. Handling the case when the container is hanging around is a great idea 🙂

I think the code you linked looks great! The only change I would suggest would be to simplify the `delete_container` method just a bit -- more specifically, to simplify the `subprocess.Popen()` call since we're just running a simple command:
```python
def delete_container() -> None:
    """Delete container if it remained"""
    proc = subprocess.Popen(shlex.split(f"docker rm -f {CONTAINER_NAME}"))
    output, stderr = proc.communicate()
    return None
```
The `proc.communicate()` call is still important here; when I was testing this locally, it seems that deleting the container may take a second to register and the `docker run` command can get stuck if the deletion wasn't waited on until completion (which `proc.communicate()` does for us).